---
title: "MARVEL"
author: "Sean Wen"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{MARVEL}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<style type="text/css">

body, td {
   font-size: 14px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

# 1. Introduction
Alternative splicing represents an additional and under-appreciated layer of complexity underlying gene expression profiles. More recently, technological advances in library preparation methodologies enabled capturing and amplification of full-length cDNAs from single cells. Thus, paving the way for splicing analysis at single-cell resolution.  

Nevertheless, single-cell splicing analysis comes with its own set of challenges including, but not limited to, low coverage of lowly-expressed genes, low capture rate of cDNA molecules, and amplification bias (*Wen et al., 2020*). To date, there remains a paucity of peer-reviewed softwares available for single-cell splicing analysis. Notable examples are BRIE (*Huang & Sanguinetti, 2017*) and Expedition (*Song et al., 2017*).  

Here, we introduce MARVEL (*M*ultimodal *A*pproach to **R**e**VE**a**L** alternative splicing dynamics at single-cell resolution). MARVEL includes features that complement existing softwares in order to more comprehensively describe and reveal splicing dynamics at single-cell resolution.

# 2. Additional Resources
Tutorials for both plate- and droplet-based alternative splicing analysis using MARVEL can be found here: https://github.com/wenweixiong/MARVEL

# 3. Additional dependencies
After loading MARVEL package, please (install and) load additional dependencies below before proceeding.
```{r, eval=FALSE}
# Load packages
library(MARVEL)

# Load additional dependencies
library(BSgenome)
library(BSgenome.Hsapiens.NCBI.GRCh38)
```
